### PR TITLE
[Dialogs] Add Dynamic Type support to Alert controller

### PR DESCRIPTION
**Pull request by ianegordon**
This adds automatic font adjustment / Dynamic Type support to MDCAlertController.

The main functionality is exposed through the mdc_adjustsFontForContentSizeCategory property.

This property is modeled after the adjustsFontForContentSizeCategory property in the UIConnectSizeCategoryAdjusting protocol added by Apple in iOS 10.0.

An example of an alert supporting dynamic type has been added to the catalog under dialogs -> alert controller.

### CODE REVIEW

**Comment by ianegordon** [comment]
The PR is back in the Travis Queue.  I believe the initial CI failure was general flakiness.

Small -> Big
screen shot 2017-02-16 at 13 37 30
screen shot 2017-02-16 at 13 37 40

File: components/Buttons/src/MDCButton.m [hunk 1]
```diff
@@ -705,6 +705,7 @@ - (void)mdc_setAdjustsFontForContentSizeCategory:(BOOL)adjusts {
 - (void)contentSizeCategoryDidChange:(NSNotification *)notification {
   UIFont *font = [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleButton];
   self.titleLabel.font = font;
+  [self sizeToFit];
```
Comment (randallli) [inline components/Buttons/src/MDCButton.m:708]: i thought the label already did this?
(action not required)

File: components/Dialogs/src/MDCAlertController.m [hunk 2]
```diff
@@ -186,6 +189,48 @@ - (void)addAction:(MDCAlertAction *)action {
...
+
+// Update the fonts used based on mdc_preferredFontForMaterialTextStyle and recalculate the
+// preferred content size.
+- (void)updateFontsForDynamicType {
+  self.titleLabel.font = [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleTitle];
+  self.messageLabel.font = [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleBody1];
+
+  // The action MDCButtons handle their own resizing
+
+  self.preferredContentSize = [self calculatePreferredContentSizeForBounds:CGRectInfinite.size];
```
Comment (randallli) [inline components/Dialogs/src/MDCAlertController.m]: we presume that this will call a layoutSubviews?
(action not required)

File: components/Dialogs/src/MDCAlertController.m [hunk 3]
```diff
@@ -186,6 +189,48 @@ - (void)addAction:(MDCAlertAction *)action {
...
+
+// Handles UIContentSizeCategoryDidChangeNotifications
+- (void)contentSizeCategoryDidChange:(NSNotification *)notification {
+  [self updateFontsForDynamicType];
+}
+
+// Update the fonts used based on mdc_preferredFontForMaterialTextStyle and recalculate the
+// preferred content size.
+- (void)updateFontsForDynamicType {
+  self.titleLabel.font = [UIFont mdc_preferredFontForMaterialTextStyle:MDCFontTextStyleTitle];
```
Comment (randallli) [inline components/Dialogs/src/MDCAlertController.m:225]: Lets see if we can use a delegation instead of inheritance. If a client would like to setup their own preferred fonts they would have to subclass and override our logic here.
How we could do it:
Instead of a bool that dictates behavior we could have a dynamicTypeDelegate property that mutates the various label's fonts of the MDCAlertController. What do you think?

**Review by randallli — CHANGES_REQUESTED** [review]

File: components/Dialogs/src/MDCAlertController.m (hunk 2 above)
Comment (ianegordon) [reply components/Dialogs/src/MDCAlertController.m]: Yes,.... eventually,... but perhaps not how you initially expect.

Our UIPresentationController listens for changes to it's presented controllers preferred content size.  This is baked in to UIPresentationController via the preferredContentsSizeDidChangeForChildContentContainer: method.  In that method our presentation controller can adjust the frame of our presented controller.  That will lead to a layoutSubviews call.

Comment (randallli) [reply components/Dialogs/src/MDCAlertController.m]: lets put a comment in that says this? mostly because of the "not how you initially expect"

File: components/Dialogs/src/MDCAlertController.m (hunk 3 above)
Comment (ianegordon) [reply components/Dialogs/src/MDCAlertController.m:225]: A big motivation for MDCAlertController was to mimic UIAlertController.  Unfortunately this includes it's hostility to subclassing "MDCAlertController class is intended to be used as-is and does not support subclassing. The view hierarchy for this class is private and must not be modified."
We may be able to expose this functionality via the UIAppearance proxy.
I've opened an issue #1215 to investigate a solution.

File: components/Buttons/src/MDCButton.m (hunk 1 above)
Comment (ianegordon) [reply components/Buttons/src/MDCButton.m:708]: Changing the font of a button will cause the button to re-layout, but it won't change the bounds.  We need to explicitly re-size the button.
In the earlier Dynamic Type PR on buttons they 'worked' as the bounds of the sample buttons handled all of the font sizes, but this was definitely a bug.

File: components/Dialogs/src/MDCAlertController.m (hunk 3 above)
Comment (randallli) [reply components/Dialogs/src/MDCAlertController.m:225]: just to clarify: are you saying we should land this as is and introduce a breaking change later? (switch from bool to delegate object)

Comment (ianegordon) [reply components/Dialogs/src/MDCAlertController.m:225]: I do believe we should land this as is.
This bool property is based off of the property added to label, text view and text field; adjustsFontForContentSizeCategory / this is the way Apple has chosen to expose it.
One of the reasons we are recommending system fonts is because of the difficulty in supporting custom fonts and dynamic type.  
If we chose to expose custom fonts here, clients will end up breaking dynamic type unless they are incredibly careful, have properly hinted fonts, and do extensive testing.
By not exposing that capability we stop all of those bugs before they start.
tl;dr: I don't believe we should support custom fonts in MDCAlertController at this time.

File: components/Dialogs/src/MDCAlertController.m (hunk 2 above)
Comment (ianegordon) [reply components/Dialogs/src/MDCAlertController.m]: Done.

File: components/Dialogs/src/MDCAlertController.m (hunk 3 above)
Comment (randallli) [reply components/Dialogs/src/MDCAlertController.m:225]: So the boolean flag is incompatible with roboto? are we checking to ensure that the client has kept the default system font? does mdc_preferredFontForMaterialTextStyle have an assert to check this?

**Review by randallli — CHANGES_REQUESTED** [review]

File: components/Dialogs/src/MDCAlertController.m (hunk 3 above)
Comment (ianegordon) [reply components/Dialogs/src/MDCAlertController.m:225]: Note that there are two similar methods for mdc_preffered*ForMaterialTextStyle.  One is on UIFont, the other is on UIFontDescriptor.

mdf_adjustsFontForContentSizeCategory is locked to system font.
This matches Apple's behavior for adjustsFontForContentSizeCategory.

UIFont mdc_preferredFontForMaterialTextStyle invoked the UIFontDescriptor methods then requests a font using that font descriptor.  The font descriptor in question does not have a font name / family specified which returns system font with the requested attributes (size, weight, etc.).

If clients wish to utilize a custom font, they can invoke UIFontDescriptor mdc_preferredFontDescriptorForMaterialTextStyle and use that descriptor as the basis to request a font with their custom font name / family.  This is likely the hook we would use if we choose to pull into RobotoLoader.  Happy to sit down and work through the details.

Comment (randallli) [reply components/Dialogs/src/MDCAlertController.m:225]: ok let me rephrase: What happens if a client sets the Typogarphy's fontLoader to the MDFRobotoFontLoader and then also turns this bool on and ask for a prefered font? They would get system font. I would expect to get a Roboto font.
This is a bug.

Comment (randallli) [reply components/Dialogs/src/MDCAlertController.m:225]: We can solve the bug by fixing the return or by preventing calls to the incompatible APIs. I would much prefer to return Roboto when setting up a Roboto font loader but if that is impossible then we should throw asserts when we try to use the APIs that would result in non consistent fonts.

**Comment by ianegordon** [comment]
Discussed Dynamic Type and our font loaders with @randallli offline.

PR #1224 will log a warning if you try and use Dynamic Type with a custom font loader.

PTAL.

**Review by randallli — APPROVED** [review]
